### ISSUE TITLE
Solus 4.5 Xfce Installed on Laptop Doesn't Provide Bluetooth

### ISSUE THREAD

**Issue by WetGeek** [getsolus/packages#1046]
### Summary

The new laptop I bought for testing on hardware has arrived, and Windows 10 has been replaced by Solus 4.5 Xfce RC2.  Everything has gone according to plan, except for one bit of drama.
![image](https://github.com/getsolus/packages/assets/6805994/e87ceaba-bfea-41ff-a0d1-b2b716d62a6b)

I cannot connect my Bluetooth mouse and keyboard to this machine because there is no Bluetooth applet in the panel, and I can't find a Bluetooth setting anywhere.

I know that the laptop offers Bluetooth.  I have two other Dell 5580 Latitude laptops like this one, so I understand it well.  It uses an Intel radio that combines Bluetooth and Wi-fi, and the BIOS is set to allow both.  That should be all that's needed.

### Steps to reproduce

1.  It would be best if this could be verified on a Latitude, but if not, perhaps another laptop with known working Bluetooth would do.
2. I completed the Windows setup I couldn't avoid.  Then I emptied the SSD using GParted, and created a 1 GiB boot partition and a 24 GiB swap partition.  The rest of the 500 GB drive was used for the root partition.
3. I installed the .ISO without any difficulties, and removed the flash drive from the USB port.
4. I configured the machine exactly the same as my testing VM, which is the same as my daily-driver laptop.
5. As part of the installation I looked for a way to connect my Bluetooth mouse and keyboard, but could not find a way.

### Expected result

I expected to find a Bluetooth icon on the panel, or at least a Bluetooth setting somewhere that I could turn on and pair with these peripherals.  I looked very carefully, but if anyone with more experience with Xfce on hardware knows what I'm missing, please let me know.

### Actual result

I was unable to find a way to use Bluetooth.

NOTE:  I have two other DELL 5580 laptops that have no problem using Bluetooth.  They both use Plasma as their DE.  And this laptop has no problem connecting to our 5G wi-fi network.  As I mentioned above, both wi-fi and Bluetooth are enabled in the BIOS settings.

### Environment

- [X] Is system up to date?

### Repo

Shannon (stable)

### Desktop Environment

Xfce

### System details

![image](https://github.com/getsolus/packages/assets/6805994/1c7be152-6c6e-4897-b5b4-98bf019443a4)

**Comment by WetGeek** [comment]
This is a particularly nasty problem for me.  This laptop is rather large, and it normally occupies a laptop stand in front of my chair.  Thus the Bluetooth peripherals make it more comfortable for me to use it, as I don't need to reach out a couple feet to access the thumbpad and keyboard on the laptop itself.  Plus, it heats up my legs when it's on my lap.  I hope we can sort out this one, because it affects me in a significant way.

**Comment by EbonJaeger** [comment]
Hm. XFCE doesn't have its own Bluetooth thing, so we'll probably have to add Blueman to the default installation.

**Comment by WetGeek** [comment]
I've been trying to figure out why I can pair the mouse with Blueman, but not the keyboard.  I know that the keyboard is using BT 5.1, as on my Plasma laptop it is simply identifid as "5.1 Keyboard" during pairing.  The mouse, which does pair with Blueman, is identified by its model name (Triathlon), and the Logitech documentation doesn't list the BT level it supports, but it's possible that it's only 5.0.  That documentation just describes its compatibility as including "Linux® kernel 2.6+", which means it should work for us.

**Comment by WetGeek** [comment]
I tried adding gnome-bluetooth using eopkg, and adding that worked without a problem.  However, it doesn't show up in the Settings menu, nor can anything it provides be run in the terminal.  Since the Xfce DE is based on GNOME, I thought there was a chance this sub-system would work with Xfce, but I haven't found a way to do that.

**Comment by WetGeek** [comment]
This is just an update.  I can see that Blueman has been added as of RC 4, but I should point out that it doesn't work any better now than it did when I tried it and reported that it couldn't scan 3 different Bluetooth keyboards that ALL work well with a 4.4 version of Solus.  To be clear, it's not a failure to _pair_ with these keyboards, it's a failure to even _see_ them at 2' from the computer.

**Comment by WetGeek** [comment]
To summarize:  It turns out that none of the Bluetooth keyboards showed up when scanned, because Blueman was set to ignore unnamed devices.  I hadn't realized that these three keyboards were all unnamed, but when that setting was turned off, they all showed up when scanned.  And the two that I tried to pair with the Xfce laptop (Logitech and generic) both paired, and connected.  But neither would actually WORK.  Typing on them, nothing would show up on the laptop.

It's significant that both of these Bluetooth keyboards have been used with an identical Latitude 5580 laptop (my "daily driver") and neither one ever had any problems there.  But on this 4.5 Xfce laptop, neither works.

**Comment by WetGeek** [comment]
TraceyC, I don't know why I haven't mentioned this before, but if no one has the resources to confirm this issue, I would be happy to make an additional one-time contribution to Solus development of an Arteck multi-device keyboard.  That's like the one I like best of the three Bluetooth keyboards that I haven't been able to use with Blueman on my Xfce laptop.  I can have one sent from Amazon, but I'd like to avoid customs complications, so an address in the US would be best.

If this offer is acceptable, you can contact me on Telegram to make arrangements privately, user name WetGeek.  Just let me know here whom I can expect to contact me that way.  If no one has Telegram available (pretty hard to believe), I can provide an email address where I can be reached as well, but a Telegram message is the far more private option.

**Comment by WetGeek** [comment]
The first test returns a failure to find the command, as it does on my other Latitude laptop, which is currently using a Bluuetooth mouse and keyboard.

The lsusb command (the second test at that link) returns nothing, given the two grep calls.  On my other Latitude laptop that call returns both values as specified in the test.

Since that appears to identify a failure,  I'm going to take a break to fix lunch, but I'll soon be back here in case I can help more.

**Comment by WetGeek** [comment]
Before I do that, I should also make something else clear.  I've been able to determine that neither of the Bluetooth keyboards I tried will work, although they can now be seen, paired, and connected.  Because I have only one Triathlon mouse, and it's in use on my main laptop, I can only say that it's seen by the Xfce laptop, but I have no idea whether it will pair, connect, or work with that device.  I suspect it will not, as the keyboards do not.

If it becomes an issue, I'm willing to plug in a wireless mouse to my main laptop, and have it "forget" the Triathlon, so I can further test the Bluetooth mouse on the Xfce laptop.  I'll be back within a very short while.

**Comment by WetGeek** [comment]
Plasma.log:
https://gist.github.com/WetGeek/c4b97de4b9b3f4e7cb95338a818d16da

**Comment by WetGeek** [comment]
Xfce.log:
https://gist.github.com/WetGeek/1af6a0859d87ec4f8683b6e5def29275

**Comment by WetGeek** [comment]
Second attempt, Plasma.log
https://gist.github.com/WetGeek/9b650fd2f6d83b8b7771b88611e50249

**Comment by WetGeek** [comment]
Second attempt, Xfce.log
(https://gist.github.com/WetGeek/2194e05568cade1468e0c3e55e30f747)

**Comment by WetGeek** [comment]
Most recent Plasma.log file:
https://gist.github.com/WetGeek/fb9b324b23446b189df202a52ec3050c

**Comment by WetGeek** [comment]
Most recent Xfce.log file:
https://gist.github.com/WetGeek/0554e1f45ce33fd0104807856eb05478

**Comment by WetGeek** [comment]
The above two files were created by first causing the Bluetooth function on both laptops to "forget" the 5.1 keyboard, then to restart bluetooth and attempt to connect the keyboard.  On the Plasma laptop, the keyboard connected and was usable.  On the Xfce laptop, the keyboard was reported as connected, but was not usable.

**Comment by WetGeek** [comment]
btmon-xfce.log
https://gist.github.com/WetGeek/abd1c29ed72db788fe3aa7ebdf116eaf

**Comment by WetGeek** [comment]
btmon on Plasma laptop:
https://gist.github.com/WetGeek/aaaeea2fdd727fc124b19e6f3fcaa0ca

**Comment by ermo** [comment]
@WetGeek 

Would you be open to trying the steps [here](https://wiki.archlinux.org/title/Bluetooth_keyboard)?

When doing this on XFCE (you don't need to attempt it on KDE), you will want to capture the btmon log in a separate terminal so you can begin spotting the point in the process where comes undone and leave the keyboard paired, but not working.

**Comment by WetGeek** [comment]
Of course I'll be glad to help.  However you caught me at breakfast, and I'm still caregiver for my wife, so it's going to be a while before I can get dressed and work on this.

Meanwhile, I have a question.  I've read "Login to the affected computer by a wired keyboard ..." and I wonder why that would be necessary. I understand that logging on with a Bluetooth keyboard would cause it to quit working as soon as Bluetooth is interrupted, but a laptop essentially has a "wired" keyboard built in.  And I've been using a wireless mouse instead of a Bluetooth for these other diagnostic exercises.  And I'm not set up here for using one computer to be a terminal to another computer, if that's what they're getting at.  Do you think this really matters?

As soon as I can, I'll re-read this document and then try to follow its recommendations to get to the bottom of this.

**Comment by WetGeek** [comment]
This is a link to a log that reprents the results of my work as documented in the Development room today, January 4.  
https://gist.github.com/WetGeek/3b231fb55fa0e7370814ff6181ebf420

**Comment by TraceyC77** [comment]
> TraceyC, I don't know why I haven't mentioned this before, but if no one has the resources to confirm this issue, I would be happy to make an additional one-time contribution to Solus development of an Arteck multi-device keyboard. That's like the one I like best of the three Bluetooth keyboards that I haven't been able to use with Blueman on my Xfce laptop. I can have one sent from Amazon, but I'd like to avoid customs complications, so an address in the US would be best.

@WetGeek , thank you again for all the troubleshooting you've done with us so far. Thank you also for this  generous offer. The Solus team has decided to get the Arteck keyboard. I've volunteered to do the testing with it. This will be for this issue, as well as with the other DEs with this ISO and future ones to ensure future reliability. The easiest way for us to handle this with OpenCollective is for me to purchase the keyboard and then Solus will reimburse me through OpenCollective.

Please share a link to the keyboard on Amazon so I know the exact model to get. I'll do the actual purchase and reimbursement.

**Comment by WetGeek** [comment]
https://www.amazon.com/Arteck-Universal-Multi-Device-Bluetooth-Smartphone/dp/B0C2VJRNRY/ref=sr_1_3?crid=RYNXI22UMKEY&keywords=arteck+hb305&qid=1704420791&sprefix=arteck+hb305%2Caps%2C141&sr=8-3

![image](https://github.com/getsolus/packages/assets/6805994/48faf06a-0b5c-4a80-9d65-6dcf534afbde)

You don't want me to just send you one, eh?

**Comment by ermo** [comment]
> You don't want me to just send you one, eh?

Well, given you are a OC contributor at the Crew level, you kind of already sponsored it? And thank you for trusting your money with us like that btw.

Tracey put up a reimbursement request for the keyboard on OC, which I just approved earlier today, so we're all set now. :sunglasses: :+1:

**Comment by TraceyC77** [comment]
After testing this with the bluetooth keyboard on an XFCE native install (Lenovo Flex) I was able to reproduce the issue, and get additional detail. Further research discovered this looks identical to a [kernel bug that had a patch submitted that unfortunately does not fix the issue for all devices](https://bugs.archlinux.org/task/62776.html)
[Related bug in the Arch bug tracker](https://bugs.archlinux.org/task/62776.html)
The bug is that the keyboard is disconnected after connecting & pairing.

@ReillyBrogan , would you mind taking a look, since this seems kernel related?

**Workaround**
I noticed that after pairing (regardless of trusting or not trusting the device), I received a system notification that the keyboard was connected, and another one saying it was disconnected. 
I was able to get the keyboard to work simply by right clicking its entry in the Bluetooth control panel and clicking Connect.

**Details**

1. Start with bluetooth keyboard not paired or trusted.
2. Open bluetooth settings.
3. Make bluetooth discoverable for 3 minutes (Adapter -> Preferences).
4. Put keyboard in pairing mode (long press bluetooth key) and click Search.
5. Select the keyboard and click the key to pair - sys notification that keyboard is Connected, then Disconnected.
6. Right click keyboard and click Connect.
7. Click check to trust (or right click keyboard and click Trust).
8. Test typing in Mousepad.

Result: successful

I followed journalctl while pairing the keyboard to see what was output at the moment the keyboard disconnected. Nothing was output during pairing, connecting and disconnecting. Nothing for dmesg either. btmon was more helpful, log attached.


[bt_keybd_connect_btmon.txt](https://github.com/getsolus/packages/files/13933290/bt_keybd_connect_btmon.txt)

Also note the large amount of these lines spammed (likely a separate issue, but noting):
```
blueman-manager[4447]: @ RAW Open: blueman-ma.. version 2.22  {0x0002} 1.604285
blueman-manager[4447]: @ RAW Close: blueman-manager           {0x0002} 1.604313
blueman-manager[4447]: @ RAW Open: blueman-ma.. version 2.22  {0x0002} 2.605634
blueman-manager[4447]: @ RAW Close: blueman-manager           {0x0002} 2.605656
```

**Comment by WetGeek** [comment]
@TraceyC77 : I was hopeful that the workaround you mentioned would work for me, so I could begin using this Xfce laptop as a daily driver, but sadly it didn't work for me.  At no time did I get that second notification that the keyboard was disconnected, just the first one that indicated it was connected.  And when I opened the context menu for the keyboard, it indicated that it was connected, yet it had no effect.

So I disconnected from the context menu, afterwhich "Connect" was once more available.  Then I clicked to connect again, and that resulted in this error:
![image](https://github.com/getsolus/packages/assets/6805994/3957255d-9296-4ca6-9e2b-8437d343aca4)

**Comment by TraceyC77** [comment]
That points to some weird difference between my Xfce installation and yours. I'd recommend checking for broken packages outlined at the top of the [Troubleshooting doc](https://help.getsol.us/docs/user/troubleshooting/)

**Comment by WetGeek** [comment]
TraceyC77, with no response to my follow-up in email, I thought I should provide an answer here, as well.  Checking two different ways for broken packages (just to be sure) revealed that there are none.  I created a new installer flash drive with the released (not RC) .ISO file, and reinstalled Xfce on my spare laptop.  Then I followed your test plan 100%, and the results were exactly as described above, including the red error message at the bottom.  Our current build of Xfce simply isn't working well with Bluetooth.

The only difference between your experience and mine seems to be our hardware.  Mine is as posted in the Summary above - DELL 5580 Latitude, all Intel where it counts.  That laptop has no problem pairing, connecting, and working with a Bluetooth mouse, but not a Bluetooth keyboard.  Your experience differs, so I assume your hardware does as well.

I'm going to try one more experiment.  Although we both used the same Bluetooth keyboard in our tests, it's possible that a simpler Buetooth mouse and keyboard, such as a typical user might have, will work where these 3-way multi-device keyboards did not, so I'm going to buy one each of those to test with.  That way I can also keep my daily-driver laptop undisturbed while I use the other one to test Xfce.  I'll let you know how that all goes.

**Comment by WetGeek** [comment]
TraceyC77:  Update.  My mouse and keyboard arrived today, and I was mistaken, or I should say, misled by the catalog description.  It's a single-device keyboard, but quite different than what I was expecting.  It provides either BT3.0, BT5.0, or wireless 2.4 (with a dongle) to a single device.

I was still happy to try something different, so I paired it using BT3.0, and got the same results as I've been getting with the various BT5.0 keyboards.  In other words, paired and connected, but not functional.  Likewise when I paired it with BT5.0.  At one point, I was able to type a few characters into Mousepad, but only a few, then that stopped, and I haven't been able to reach that state again.

One thing works, and makes it worth the $16+ I paid for this pair.  The wireless function with a dongle works just fine.  So, finally I'm able to put this on the laptop stand in front of my chair, and use it longer-term as I would my daily-driver Plasma laptop.  That's what I'm doing now.  Perhaps I'll be better able to test Xfce as other issues are resolved.

And except for this Bluetooth problem, this Xfce laptop has the identical UI and functionality as my Plasma laptop, so working with it full-time isn't much of a sacrifice.

**Comment by TraceyC77** [comment]
Sorry for the delay in my response, and thank you for your patience.

Please keep replies to any issue in the issue, rather than reply in emails. This keeps all communication in one place.

We want to make sure our kernel includes all the mods necessary for your hardware. To do this, can I ask you to do the following? The goal is to get output from your Xfce install, and then a live Manjaro install.

1. Boot into Xfce, and attempt to connect to the bluetooth keyboard (the model we both have). It doesn't matter if it succeeds or not.
2. Run this in your terminal. It will list all your kernel modules and write the output to a file in your home directory:
`lsmod | sort > ~/lsmod_output_xfce.txt`
3. Open up that file in your favorite editor and paste the output to the pastebin of your choice (I use https://bpa.st/). Save the link.
4. Download a recent version of Manjaro (https://manjaro.org/download/) (please use 23.1.3 based images). Please use either GNOME or Plasma.
5. Create a bootable USB with Manjaro. Boot it on this same machine.
6. You'll need a way to save your lsmod output, so attach a USB drive to your system with Manjaro running. Determine what directory its using with `findmnt | grep sd`. (You can skip this if you prefer to copy and paste from the terminal to a file).
7. Attempt to connect your bluetooth keyboard. It doesn't matter if its successful.
8. Run `lsmod | sort > /DIRECTORY/lsmod_output_manjaro.txt` - replace "DIRECTORY" with the directory your USB drive is on. (for example, on my system that would be /run/media/tracey/docs/lsmod_output_manjaro.txt)
9. Paste the lsmod output to another https://bpa.st/ (Again, you can just copy and paste the lsmod output in terminal to the site, if you prefer).
10. Copy the two links from the pastebin site to this issue, let us know which one is which.

After we get this information, we'll see if there are mods we need to enable for your hardware. We'll let you know what the result is, and if we need you to re-test the keyboard. Thank you for helping us look into this bug.

**Comment by WetGeek** [comment]
lsmod from laptop:  https://bpa.st/UJWQ

lsmod from Manjaro live:  https://bpa.st/XI5A

**Comment by TraceyC77** [comment]
ping @ReillyBrogan

**Comment by WetGeek** [comment]
@ReillyBrogan: Is there any progress to report yet?  This one's kinda personal.

**Comment by ReillyBrogan** [comment]
There's nothing in the hardware report that we haven't enabled in our kernel already, so unfortunately this one isn't a kernel issue (probably)

**Comment by WetGeek** [comment]
When I looked at the files with kdiff3, I couldn't see anything obvious, either.  Does that mean we need to identify something other than Blueman that might be used instead?   Do you know of any candidates?  Manjaro worked fine - obviously they found something.

**Comment by ermo** [comment]
> lsmod from laptop: https://bpa.st/UJWQ
> 
> lsmod from Manjaro live: https://bpa.st/XI5A
> 
> Incidentally, when I paired the keyboard with the Manjaro "live" session, it worked fine. Also, if it matters, the Manjaro DE was KDE Plasma.

Just as a crazy, off-the-cuff suggestion, would it interest you at all to do the same thing with an Xfce-based EndeavourOS live USB key and/or VM install?

That might provide a more apples-to-apples comparison...?

**Comment by WetGeek** [comment]
I'll be glad to do anhthing that might help - as I've said, this one is personal.  And your suggestion sounds as it it qualifies, for sure.  I need to prepare lunch soon, but afterwards I'll get started on it.

**Comment by WetGeek** [comment]
I just remembered that I need to use bpa to create a url I can post here, so I'll do that now.

Here's lsmod_output_elementary:  https://bpa.st/45KPA

When I started the "live" session, there was no way for me to select a DE to download, so I assumed that Xfce is their only choice.  And this looked like a very modified Xfce to me.  Somehow I also avoid the $20 entry fee to download it.

**Comment by EbonJaeger** [comment]
Did you download elementary OS instead of Endeavour OS?

**Comment by WetGeek** [comment]
Ahhh ... yeah.  Sorry.  By the time lunch was over, I guess I'd got them confused,  Give me a little while longer, and I'll get Endeavour OS for you.

**Comment by WetGeek** [comment]
Sorry to take so long - I had some download problems getting the Endeavour ISO file.

lsmod_output_endeavour.txt:  https://bpa.st/LLXE4

**Comment by WetGeek** [comment]
Here's what's probably relevant, but I'm not sure what to look for.  There are certainly a lot of differences this time.

![image](https://github.com/getsolus/packages/assets/6805994/0b39b7ec-d5df-47b1-9b28-a11f1f1695b7)

**Comment by TraceyC77** [comment]
@WetGeek , I noticed on Plasma 6 that the keyboard connects but doesn't send any keypresses to the system **if another bluetooth device is connected**

When you were trying to pair and type on the keyboard, were any other bluetooth devices connected?

Can you try pairing the keyboard and using it with no other bluetooth device connected?

**Comment by WetGeek** [comment]
My experience was similar, but not exact.  With newer syncs, I've been able to consistently connect, but never been able to send characters to the system.  At times, I'm sure I've had a Triathlon mouse connecterd (successfully) as well, but at other times I've just been interestd in the keyboard, so attempted to use it by itself.  I didn't notice any differences regarding the success of the keyboard.

Since we've done nothing Xfce-related for a long time, my spare laptop is currently running Plasma 6.  I can reload it with Xfce again, Solus Plasma, of course.  I'll start tonight.  To be clear, you'd just like me to see whether I can reproduce your result, right?  That would be a workig connection for the keyboard with no other Bluetooth device paired.

**Comment by WetGeek** [comment]
Mischief managed.  At least far enough to confirm that I can use both the keyboard and the mouse on this Plasma 6 laptop.  Both had previous connected to Device 3 to my smartphone, so I shut off Bluetooth on the phone, and paired both keyboard and mouse to the Plasma 6 system I just finished installing.

In fact, I'm writing this on the spare laptop, using the mouse and keyboard now.  Both connected very smoothly, as I would expect on a Plasma system.  I've removed my daily laptop from its stand and replaced it with this one.  I'm considering building this one out all the way, and replacing my daily laptop with this one.  But not tonight.  I'm beat.  Gonna hibernate this laptop, then go hibernate myself.

Bottom line:  both devices paired smoothly, and work together just fine.

**Comment by WetGeek** [comment]
Upon wake-up from hibernation in the morning, both devices were still paired, but not connected.  This machine still had the dongle for the wi-fi mouse inserted, so I used that device to select the Bluetooth icon in the system tray and reconnect both Bluetooth devices.  I won't always have that second mouse connected, though, so I'm going to spend as long as it takes to see if there's a way to make them reconnect automatically on wake-up.  I noticed that there's a setting that might have something to do with that, and it's turned on by default on my Plasma 4.5 laptop, but not on the Plasma 6 laptop.  I'm going to create a new issue to deal with this matter, since this one is about Xfce, it's long, it's unfinished, and figuring out a Plasma mouse problem is off-topic here.

**Comment by TraceyC77** [comment]
This issue is about the bluetooth keyboard behavior on xfce. I need you to confirm what I asked on xfce.

On xfce, if no other bluetooth devices are paired, can the bluetooth keyboard successfully pair and send input?

If you are no longer using xfce on the device, and plan on no longer using it, feel free to close this issue out since it wasn't able to be proruced.

**Comment by WetGeek** [comment]
Closing per advice from @TraceyC77.
